# Linear Algebra 2 - Dot products and Hyperplanes

### Topics

1. Vectors
2. Norms
3. Matrix Multiplication
4. Dot product
5. Angle between two vectors
6. Unit Vectors
7. Projection of a Vector

## 1 Vector Algebra

### 1.1 Vectors

#### Definition

A quantifiable entity that has **both magnitude and direction** is called as Vector.

#### Properties

1. A Vector has only **one dimension**.
2. Mathematically, a vector $\vec{x}$ is represented as an array of numbers (a column).
3. In ML Context, each data-point (a single row in the dataset) is represented as a vector on a Cartesian plane.
4. In ML, the vector formed from a data-point is called as **Feature Vector**.

#### Vector Representation

Mathematical notation of a vector $\vec{x}$ having $d$ dimension containing real numbers:

$$
\large
\begin{aligned}
x \in \mathbb{R}^d
\end{aligned}
$$

Where $d$ represents number of elements in the vector.

> **Note**:
>
> A vector can be represented as a row vector or a column vector, but when it is not explicitly specified, a vector is **assumed to be a column vector**.

#### Row Vector

$$
\large
\begin{aligned}
\vec{x} = [x_1 \; x_2 \; \ldots \; x_d]
\end{aligned}
$$

#### Column Vector

$$
\large
\begin{aligned}
\vec{x} = \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_d \end{bmatrix}
\end{aligned}
$$

### 1.2 Magnitude

#### Definition

1. Magnitude of a point is the distance of that point from the origin represented as a **position vector**.
2. Magnitude of a vector is the shortest distance (displacement) between the origin and the point.

#### Properties

1. Shortest distance between two points is called as Displacement.
2. Displacement between two points is calculated using **Euclidean Distance**.
3. Magnitude or length of a vector is also called as **Norm** of a vector.
4. Mathematically, Norm of a vector is represented as $\Vert\bar{x}\Vert$.
5. In ML context, Norm of a 2D vector is its length from the origin $(0, 0)$ on a Cartesian plane.

### 1.3 Euclidean Distance

#### Definition

Euclidean distance is the straight-line distance (shortest distance) between two points.

#### Formula

Formula to calculate Euclidean Distance between two points $(x_1, y_1)$ and $(x_2, y_2)$

$$
\large
\begin{aligned}
d = \sqrt{(y_2 - y_1)^2 + (x_2 - x1)^2}
\end{aligned}
$$

### 1.4 Manhattan Distance

#### Definition

Manhattan distance is the total horizontal and vertical path between two points, mimicking navigation in a grid-like city layout.

#### Formula

Formula to calculate Manhattan Distance between two points $(x_1, y_1)$ and $(x_2, y_2)$

$$
\large
\begin{aligned}
d = \vert{y_2} - {y_1}\vert + \vert{x_2} - {x_1}\vert
\end{aligned}
$$

### 1.5 Types of Norms

Norms are used for data-points of higher dimensions. There are two types of Norms of a vector:

1. L1 Norm
2. L2 Norm

There are many more techniques to calculate distance of a vector but L1 and L2 are more popular.

#### L2 Norm of a Vector

Formula for L2 Norm or Euclidean Distance of a vector of $d$ dimension:

$$
\large
\begin{aligned}
{\Vert\vec{x}\Vert}_2 &= \sqrt{{x_1}^2 + {x_2}^2 + {x_3}^2 + \ldots + {x_d}^2} \\[12pt]
                      &= \sqrt{\sum_{i=1}^d {x_i}^2}
\end{aligned}
$$

#### L1 Norm of a Vector

Formula for L1 Norm or Manhattan Distance of a vector of $d$ dimension:

$$
\large
\begin{aligned}
{\Vert\vec{x}\Vert}_1 &= \vert{x_1}\vert + \vert{x_2}\vert + \vert{x_3}\vert + \ldots + \vert{x_d}\vert \\[12pt]
                      &= \sum_{i=1}^d \vert{x_i}\vert
\end{aligned}
$$

> **Note**:
>
> When Norm of a Vector $\Vert{x}\Vert$ is not explicitly specified, it is **assumed to be L2 Norm** by default.

### 1.6 Direction

Direction of a Vector is determined by its angle to the x-axis.

## 2 Matrix Multiplication

### 2.1 Multiplication Criteria

Lets say we have two matrices $A: m \times n$ and $B: n \times k$,

$$
\large
\begin{aligned}
m \times n \Leftrightarrow n \times k = m \times k
\end{aligned}
$$

1. Number of columns in the first matrix should be equal to the number of rows in the second matrix.
2. Matrix multiplication is not commutative, i.e., $A \times B \ne B \times A$

https://www.geogebra.org/m/ETHXK756

### 2.2 Dot Product

#### Properties

1. Dot product calculates the **linear similarity** between two vectors.
2. Dot product is specifically used between two vectors.
3. Higher the value of dot product higher is the similarity.
4. Lower the value of dot product lower is the similarity.

#### Formula

Formula to calculate dot product of two vectors $\vec{a}$ and $\vec{b}$:

$$
\large
\begin{aligned}
\vec{a} \cdot \vec{b} &= a^T \cdot b  \\[12pt]
                      &= \sum_{i=1}^d a_i b_i
\end{aligned}
$$

### 2.3 Examples

In [1]:
import numpy as np

#### Example #1: Matrix Multiplication

##### Case #1: Same shape

In [2]:
m1 = np.array(
    [
        [1, 2],
        [3, 4],
    ]
)  # 2 X 2
m2 = np.array(
    [
        [5, 6],
        [7, 8],
    ]
)  # 2 X 2

In [3]:
[
    [(1 * 5 + 2 * 7), (1 * 6 + 2 * 8)],
    [(3 * 5 + 4 * 7), (3 * 6 + 4 * 8)],
]  # 2 X 2

[[19, 22], [43, 50]]

or

In [4]:
m1 @ m2

array([[19, 22],
       [43, 50]])

or

In [5]:
np.matmul(m1, m2)

array([[19, 22],
       [43, 50]])

##### Case #2: Different shape

In [6]:
m1 = np.array(
    [
        [1, 2, 3],
        [4, 5, 6],
    ]
)  # 2 X 3
m2 = np.array(
    [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 0, 1, 2],
    ]
)  # 3 X 4

In [7]:
[
    [(1 * 1 + 2 * 5 + 3 * 9), (1 * 2 + 2 * 6 + 3 * 0), (1 * 3 + 2 * 7 + 3 * 1), (1 * 4 + 2 * 8 + 3 * 2)],
    [(4 * 1 + 5 * 5 + 6 * 9), (4 * 2 + 5 * 6 + 6 * 0), (4 * 3 + 5 * 7 + 6 * 1), (4 * 4 + 5 * 8 + 6 * 2)],
]  # 2 X 4

[[38, 14, 20, 26], [83, 38, 53, 68]]

or

In [8]:
m1 @ m2

array([[38, 14, 20, 26],
       [83, 38, 53, 68]])

or

In [9]:
np.matmul(m1, m2)

array([[38, 14, 20, 26],
       [83, 38, 53, 68]])

#### Example #2: Two Vectors

In [10]:
# Vector Dot Product
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

In [11]:
[1 * 4 + 2 * 5 + 3 * 6]

[32]

or

In [12]:
np.dot(v1, v2)

np.int64(32)

In [13]:
np.matmul(v1, v2)

np.int64(32)

#### `np.dot` vs `np.matmul`

In [14]:
# Source - https://stackoverflow.com/q/34142485

a = np.random.rand(8, 13, 13)
b = np.random.rand(8, 13, 13)

In [15]:
np.matmul(a, b).shape

(8, 13, 13)

In [16]:
np.dot(a, b).shape

(8, 13, 8, 13)

> **Note**:
>
> `np.dot()` and `np.matmul` are **NOT** same. They behave differently when data is of higher dimensions > 2.

### 2.4 Usecase

#### Image Similarity

1. Images are converted to vectors (real numbers).
2. Similar images will have higher dot product values.
3. Dissimilar vectors will have lower dot product values.

#### Text Similarity

1. Texts are converted to vectors (real numbers).
2. Similar images will have higher dot product values.
3. Dissimilar vectors will have lower dot product values.

> **Note**:
>
> Computers can process only numbers hence all types of data (text, images, etc.) needs to be converted to numbers before building a ML model.

## 3 Applications of Dot Product

### 3.1 Angle between Vectors

#### Definition

Angle between the vector of the data-point and positive x-axis.

#### Formula

Formula to calculate angle between two vectors $\vec{a}$ and $\vec{b}$

$$
\large
\begin{aligned}
\cos \uptheta = \frac{a^T \cdot b}{\Vert a \Vert \; \Vert b \Vert}
\end{aligned}
$$

In [17]:
angles = [0, 90, 180, 270, 360]

for idx in range(len(angles) - 1):
    a1 = angles[idx] + 1 if angles[idx] else 0
    a2 = angles[idx + 1]
    c1 = np.cos(np.radians(a1)).round(1)
    c2 = np.cos(np.radians(a2)).round(1)

    print(f"[{a1:>3}, {a2:>3}] = {c1:>4}, {c2:>4}")

[  0,  90] =  1.0,  0.0
[ 91, 180] = -0.0, -1.0
[181, 270] = -1.0, -0.0
[271, 360] =  0.0,  1.0


<div style="display: inline-block">

| Quadrant |      Angle | $Cos\;\uptheta$ |
| -------: | ---------: | :-------------- |
|        1 |    0 to 90 | Positive        |
|        2 |  91 to 180 | Negative        |
|        3 | 181 to 270 | Negative        |
|        4 | 271 to 360 | Positive        |

</div>

#### Properties

1. If dot product between two vectors is zero then they are perpendicular to each other because $\cos 90^{\circ} = 0$.

#### Usecases

Cosine Similarity

### 3.2 Machine Learning

In ML context, the dot product is used to represent the below "best fit" line, hyper-plane equations:

$$
\large
\begin{aligned}
w_1 x_1 + w_2 x_2 &+ w_0 = 0 \\[5pt]
\text{or} & \\[5pt]
w_1 x_1 + w_2 x_2 + \ldots +  w_d x_d &+ w_0 = 0
\end{aligned}
$$

using $\vec{w}$ and $\vec{x}$ vectors:

$$
\large
\begin{aligned}
\vec{w} = \begin{bmatrix} w_1 \\ w_2 \\ \vdots \\ w_d \end{bmatrix}
\qquad
\vec{x} = \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_d \end{bmatrix}
\end{aligned}
$$

as,

$$
\large
\begin{aligned}
{w^T}\cdot{x} &+ w_0 = 0
\end{aligned}
$$

Where ${w^T}\cdot{x}$ is the dot product between weight vector $\vec{w}$ and feature vector $\vec{x}$.

#### Properties

1. Intercept $w_0$ is not included in $\vec{w}$ vector.
2. Weight vector: $\vec{w}$ is always perpendicular to Decision Boundary, hence it is also called as **Normal Vector**.

#### Example

1. There is a Dataset $X$ with 250 data-points and 4 features i.e., 250 rows and 4 columns.
2. $X$ is of shape: [4, 250] (250 vectors horizontally stacked to form $X$).
3. A weight vector $\vec{w}$ of shape [4, 1] i.e., 4 rows and 1 columns was obtained after optimization.
4. $w^T$ will be of shape: [1, 4]
5. The resulting Decision Boundary (ML model) will be of shape [1, 250] i.e., $1 \times 4 \Leftrightarrow 4 \times 250 = 1 \times 250$
6. ${w^T}\cdot{x} = 1 \times 250$

### 3.2 Unit Vector

#### Definition

A vector whose magnitude equals to one.

$$
\large
\begin{aligned}
\Vert a \Vert = 1
\end{aligned}
$$

#### Formula

Formula to calculate unit vector $\widehat{w}$ of a $\vec{w}$ vector:

$$
\large
\begin{aligned}
\widehat{w} = \frac{\vec{w}}{\Vert{w}\Vert}
\end{aligned}
$$

#### Properties

1. Unit vector is used to identify the **direction of a vector**.

### 3.3 Projection of a Vector

#### Definition

$$
\large
\begin{aligned}
\text{Projection of $\vec{x}$ on $\vec{y}$} = \vec{x}\cdot\widehat{y}
\end{aligned}
$$

#### Formula

By substituting the formula for unit vector $\widehat{y}$ in above equation we get:

Formula to calculate projection of $\vec{x}$ on $\vec{y}$:

$$
\large
\begin{aligned}
\text{Projection of $\vec{x}$ on $\vec{y}$} = \frac{\vec{x}\cdot\vec{y}}{\Vert{y}\Vert}
\end{aligned}
$$

#### Properties

1. If projection of $\vec{x}$ on $\vec{y}$ is zero, because their dot product is zero, then $\vec{x}$ and $\vec{y}$ vectors are perpendicular.

#### Usecases

1. Projection of a vector is used in Principal Component Analysis (PCA).
2. Projection is used in feature extraction.